Lib/subprocess.py

```diff
@@ -796,6 +801,22 @@ def __exit__(self, type, value, traceback):
             if self.stdin:
                 self.stdin.close()
         finally:
+            if exc_type == KeyboardInterrupt:
+                # https://bugs.python.org/issue25942
+                # In the case of a KeyboardInterrupt we assume the SIGINT
+                # was also already sent to our child processes.  We can't
+                # block indefinitely as that is not user friendly.
+                # If we have not already waited a brief amount of time in
+                # an interrupted .wait() or .communicate() call, do so here
+                # for consistency.
+                if self._sigint_wait_secs > 0:
```

Comment:
Since this code is repeated is several places (`wait`, `communicate`), how about factoring it out?

Reply:
__exit__ and the Popen instance have no knowledge of timeouts or whether the child process should be killed and in what circumstances.  I didn't want to move that into the Popen instance.

If I do, it would be internal-only APIs.  I don't want to complicate the public Popen API surface further.

Reply:
You could simply have a `self._handle_keyboard_interrupt(timeout)` helper doing the `_sigint_wait_secs` dance. You would also centralize the comment there.